### ISSUE TITLE
[Case Study] When yarn node program restarts itself, the new process lacks yarn context

### ISSUE THREAD

**Issue by AlexLeung** [yarnpkg/berry#1218]
**What package is covered by this investigations?**

`yarnpkg-cli`

**Describe the goal of the investigation**

Allow a `yarn node` process to call a CLI stored in .yarn PNP zips, even after this `yarn node` process has been restarted via itself.

**Investigation report**

I've created a demo repository: https://github.com/AlexLeung/yarn-node-restart-demo
After cloning run `yarn demo`

Basically, the script will spawn `tsc -b --watch`, wait 10 seconds, then the script will kill the active tsc child process then restart itself via

```typescript
process.on("exit", function () {
    require("child_process").spawn(process.argv.shift(), process.argv, {
        cwd: process.cwd(),
        detached: true,
        stdio: "inherit"
    });
});
process.exit();
``` 
The new process will ENOENT when trying to spawn `tsc` again.

Another approach I tried was changing 
- from `tsc -b --watch`
- to `yarn tsc -b --watch`

this allows the restarted process to spawn tsc, however the original `tsc -b --watch` will not die, I think because `yarn tsc -b --watch` will create several other detached processes that aren't killed when the original is. You can tell this is happening because if you save a file after a few restarts, you'll see the same tsc message being logged multiple times.

I think the solution would be for yarn to make sure all child processes are killed upon receiving a SIGTERM.

**Comment by arcanis** [comment]
I don't think we'll change that - basically what happens is that Yarn creates a temporary directory when spawning commands. We put inside this folders all the binaries that the command can access, and store the directory path into the PATH environment variable.

To prevent the disk space from growing needlessly, the temporary folder is erased as soon as the command returns. Since in your case you detach the process, the folder is cleared even though you're still executing. There's little we can do here apart not clearing the folder, which isn't an acceptable solution.

**Comment by AlexLeung** [comment]
I see. Should I open an issue for that alternative approach I mentioned then?

The one where when using `yarn tsc -b --watch`, a SIGTERM sent to the spawned child_process only terminates the yarn process, but not the actual tsc process.

**Comment by AlexLeung** [comment]
I found for my purposes (having a node program which can restart itself), I was able to achieve this without multiple processes by using node Workers.

On 

```
if(require('worker_threads').isMainThread)
```

I create a child thread

```
new require('worker_threads').Worker(__filename);
```

which will keep the pid and process args the same. On restart, the child worker can send a message to the parent and the parent will then know to restart the child worker when the child kills itself.
